### ISSUE TITLE
Unable to build in devcontainer

### ISSUE THREAD

**Issue by QuentinLCS** [expressive-code/expressive-code#224]
### What version of `astro-expressive-code` are you using?

0.35.3

### What version of `astro` are you using?

4.11.5

### What package manager are you using?

pnpm ( seems to be the same with npm )

### What operating system are you using?

Mac

### Describe the Bug

When I run astro ( `pnpm run dev` ) in my devcontainer, the build freezes without throwing any error. 

Note : It works as soon as I remove `astro-expressive-code` from the dependencies. 

### Link to Minimal Reproducible Example

https://github.com/QuentinLCS/astro-starlight

**Comment by hippotastic** [comment]
Thank you for the report. I unfortunately have no idea what devcontainers are and never used them, so I don't have an idea how to even test this. Is there any more information you can provide or is there anyone familiar with devcontainers to help?

The log.txt in your example does not seem to match your code. The code does not use Starlight (commented out), but the log shows Starlight.

Just to be sure: There is no need to add both Starlight and Expressive Code in your Astro config. Starlight already comes with Expressive Code built in.

**Comment by QuentinLCS** [comment]
You're right, the log.txt can be ignored. I forgot to delete it. I first thought that it was a Starlight error, my bad !

By installing the vscode recommended extension, you should be able to run the workspace in a docker container.

Once started, running Astro with the dependency will result in a freeze.

**Comment by hippotastic** [comment]
I don't use Docker, so I cannot reproduce this myself. Is there any sort of log available that allows you to see why it's freezing? Is this also freezing on a regular fresh Starlight project without any changes?

**Comment by QuentinLCS** [comment]
> Is there any sort of log available that allows you to see why it's freezing?

I have 0 logs running `pnpm run dev --verbose`

>  Is this also freezing on a regular fresh Starlight project without any changes?

It is

**Comment by hippotastic** [comment]
Hmm. Could you be hitting some sort of default memory limit in your devcontainer? Maybe there are devcontainer config options you could increase?

If it's a memory issue, the only part of EC that uses a lot of memory is Shiki. You could try disabling the default Shiki integration by passing `shiki: false` in the EC configuration options and see if that helps.

**Comment by delucis** [comment]
IIUC you can also test a devcontainer using GitHub CodeSpaces right?

I tried opening https://github.com/QuentinLCS/astro-starlight in a CodeSpace and it seemed to work fine, but maybe I’m doing something wrong. (Had to update an import in `src/content/config.ts` that was still loading Starlight although not installed, but apart from that the dev server seemed to be working.)

**Comment by QuentinLCS** [comment]
> I tried opening https://github.com/QuentinLCS/astro-starlight in a CodeSpace and it seemed to work fine

I didn't try in codespace and it runs well. I don't know if codespace is built differently or if the problem occur only on apple silicon

**Comment by hippotastic** [comment]
Coming back to this issue, I just performed some reproduction attempts.

I set up Docker and the VS Code devcontainers extension on two of my machines (one running Windows 10, one running macOS Sonoma), and was able to successfully build the entire EC repo including its Starlight-powered docs site, as well as a plain Starlight starter project, both of which you stated didn't work for you.

Please note that my macOS machine is a Macbook Air with M1 chipset, so Apple Silicon is probably also not the source of your issues.

However, what I found online is that there are many open issues concerning devcontainers on macOS hosts in general (nothing specific to EC). Some users report that pnpm fails to install packages properly when reopening a workspace in a devcontainer as offered by VS Code, as pnpm fails to create hardlinks to the installed packages in this scenario. As this is something I encountered as well on macOS during my testing (but only when using pnpm), I committed a fix to EC's own devcontainer file that now runs `pnpm config set store-dir ~/.local/share/pnpm/store` before attempting to install any packages, which solves this issue. However, as you reported that the issue seems to be the same for you with npm, I also don't think that this is the cause of your problems.

Long story short, after all the testing I've done, I'm unable to reproduce your issue. Combining this with the fact that no other EC or Starlight users have reported any similar issues, I believe that the cause must be something in your own hardware or software configuration. I'm therefore closing this issue now. Happy to reopen in case there are any new findings!

**Comment by QuentinLCS** [comment]
Update : 

I observed that you used Node 18.20.4

Using this version, it works well. 

Upgrading to node 22.7.0 makes it fail as described

**Comment by hippotastic** [comment]
This sounds like an interesting lead, I'll give it a try soon and keep you posted.

**Comment by hippotastic** [comment]
I just built the entire EC repo successfully on a Windows host with the image `typescript-node:22-bullseye`, which uses Node v22.6.0 by default.

Afterwards, I ran `nvm install 22.7.0` to switch to your failing Node version, verified that it's actually being used through `node -v`, and then reran `pnpm astro build` in the `docs` folder to build the Starlight-based EC docs site again. This also completed without any errors.

In other words, sadly, the Node version alone also doesn't seem to explain the errors you're encountering on your machine. It might be useful if you could try it on a different development machine to ensure your setup isn't the cause of the issue.

**Comment by hippotastic** [comment]
...and now I can also confirm that it works fine on my Macbook Air with M1 chipset, as you mentioned Apple Silicon being a potential source of the issue earlier. I just repeated the same steps as described above there and could also build the EC repo including its Starlight docs there with Node v22.6.0 and v22.7.0.

**Comment by QuentinLCS** [comment]
Even with pnpm astro build ?

**Comment by hippotastic** [comment]
> Even with pnpm astro build ?

Yes. I followed exactly the same steps as mentioned before.

**Comment by QuentinLCS** [comment]
Weird, I'll try to use your image with node 22 as you did

**Comment by QuentinLCS** [comment]
It seems like it works with typescript-node:22-bullseye but not my custom docker image, I'm looking into it

**Comment by QuentinLCS** [comment]
No reason found, since it works with the typescript-node image, I'll customize mine base on this one. 

Thanks for your help :)

**Comment by hippotastic** [comment]
Sure thing, glad we found a solution that works for you! Happy coding :)

**Comment by QuentinLCS** [comment]
I have an hypothesis. 

I still have the issue when my Docker image is built by the GitHub Actions even with the "typescript-node". 

But it works when I create a local Dockerfile, built locally.

**Comment by hippotastic** [comment]
I have no idea, sorry! :(

Honestly, right from the beginning, this entire issue looked like not actually being caused by EC, but like some machine/platform incompatibility or issue that just happens to get randomly triggered by the presence of EC due to an unknown reason.

In other words, I doubt that this really is an issue I can debug and/or fix, and think that the root cause actually lies outside of this project. So far, it has really proven to be impossible to pinpoint and debug, although we both already invested a considerable amount of time, and I'd not like to sink even more time into this.

At this point, I'd recommend picking the simplest workaround and rolling with it. Probably just the Node version that works fine for you.

**Comment by QuentinLCS** [comment]
Alright, thanks

**Comment by QuentinLCS** [comment]
If someone has the same issue, GitHub Actions solution to build a docker image for multiple platforms : 

```yaml 
// .github/workflows/ci.yml

name: Publish Docker image

on:
  release:
    types: [published]

jobs:
  push_to_registry:
    name: Push Docker image to Docker Hub
    runs-on: ubuntu-latest
    steps:
      - name: Check out the repo
        uses: actions/checkout@v4

      - name: Set up QEMU for ARM builds
        uses: docker/setup-qemu-action@v3

      - name: Set up Docker Buildx for multi-platform builds
        uses: docker/setup-buildx-action@v3
      
      - name: Log in to Docker Hub
        uses: docker/login-action@v3.2.0
        with:
          username: ${{ secrets.DOCKER_USERNAME }}
          password: ${{ secrets.DOCKER_TOKEN }}
      
      - name: Extract metadata (tags, labels) for Docker
        id: meta
        uses: docker/metadata-action@v5.5.1
        with:
          images: IMAGE_NAME
      
      - name: Build and push Docker image
        uses: docker/build-push-action@v5.3.0
        with:
          context: .
          file: ./Dockerfile
          platforms: linux/arm64, linux/amd64
          push: true
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}
  ```

**Comment by hippotastic** [comment]
Thank you for posting the solution!

I'm glad you managed to solve it. I was way beyond my knowledge there and didn't really believe anymore that it was possible to make this work without any compromises (hence the recommendation to pick a known workaround). Actually solving this feels way more satisfying!

If this is actually a Node issue (and not a Docker one), I wonder if it might be helpful to create an issue in the Node repo. I still have no idea why the additional steps are even necessary to get an image where the latest Node release works in multiple CPU architectures, and that it just freezes without these precautions looks like a bug to me. Probably something inside Node changed or was compiled differently?
